File: lab03/cpp/CngAddTab.cpp
#include <iostream>
#include <vector>
#include <iomanip>

using namespace std;

// Function to compute GCD of two numbers
int gcd(int a, int b) {
    a = abs(a);
    b = abs(b);
    while (b) {
        int temp = b;
        b = a % b;
        a = temp;
    }
    return a;
}

// Function to generate the addition table
vector<vector<string>> generate_table(int n, int start_x, int start_y, int width, int height, bool toggle) {
    vector<string> header;
    
    header.push_back("");
    for (int k = start_x; k < start_x + width; k++) {
        header.push_back(to_string(k));
    }

    vector<vector<string>> power_table;

    for (int a = start_y; a < start_y + height; a++) {
        vector<string> current_row;
        current_row.push_back(to_string(a));

        for (int k = start_x; k < start_x + width; k++) {
            int result = (a + k) % n;
            int common_factor = gcd(result, n);

            if (!toggle || (common_factor == 1)) {
                current_row.push_back(to_string(result));
            } else {
                current_row.push_back("");
            }
        }
        power_table.push_back(current_row);
    }

    // Add header as the first row of the table
    power_table.insert(power_table.begin(), vector<string>(header.size()));
    for (size_t i = 0; i < header.size(); ++i) {
        power_table[0][i] = header[i];
    }

    return power_table;
}

// Function to print the table
void print_table(const vector<vector<string>>& table) {
    for (const auto& row : table) {
        for (const auto& item : row) {
            cout << setw(10) << item << " "; // Adjust the width as needed
        }
        cout << endl;
    }
}

int main() {
    int start_x, start_y, width, height, n;

    cout << "Enter n: ";
    cin >> n;

    vector<vector<string>> table = generate_table(n, 0, 0, n, n, false);
    print_table(table);

    return 0;
}

File: lab03/py/CngAddTab.py
from tabulate import tabulate

def generate_table(n, start_x, start_y, width, height, toggle):
    header = [k for k in range(start_x, start_x + width)]
    
    add_table = []
    
    for a in range(start_y, start_y + height):
        common_factor = gcd(a, n)
        current_row = [a]
        
        for k in range(start_x, start_x + width):
            result = (a+k) % n
            if not toggle or (common_factor == 1):
                current_row.append(result)
            else:
                current_row.append('')
        
        add_table.append(current_row)
    
    return [header, add_table]

def gcd(a, b):
    a, b = abs(a), abs(b)
    while b:
        a, b = b, a % b
    return a

def main():
    n = int(input("Enter n: "))

    headers, table = generate_table(n, 0, 0, n, n, False)
    print(tabulate(table, headers=headers, tablefmt="grid"))

main()